### DISCUSSION TITLE
Cancelling task in NET 7.0

### DISCUSSION THREAD

**Discussion by ebezak** [dotnet/runtime#93063] [Q&A]
In our unit tests we use certain behavior of CancellationToken. After upgrading from .NET 6.0 to .NET 7.0 several tests failed. The reason is change in the behavior which can be easily explained with the following code.

```
class Program
{
    static void Main()
    {
        var cts = new CancellationTokenSource();
        var task = DoAsync(cts.Token);
        cts.Cancel();
        Console.WriteLine(task.IsCanceled);
    }

    static async Task DoAsync(CancellationToken ct)
    {
        await Task.Delay(100, ct);
    }
}
```

In .NET 6.0 is the output True but in .NET 7.0 is the output False. Interestingly if I use a different task

`var task = Task.Delay(1000, cts.Token);`

the behavior is consistent across both .NET versions and outputs True.
Is this intended? Currently we use workaround of Thread.Sleep(10) before checking task.IsCanceled.

**Comment by stephentoub** [comment]
This is by design. The work item is queued rather than invoked synchronously so that all of the work after the await doesn't happen synchronously as part of the call to Cancel.

**Reply by ebezak** [reply]
I understand this is asynchronous. I was just wondering if this is intended as I couldn't find this breaking change in release notes for .NET 7.0. In .NET 6.0 the behavior is different.

**Comment by hez2010** [comment]
You can reaplace the line `Console.WriteLine(task.IsCanceled);` with `task.ContinueWith(_ => Console.WriteLine(task.IsCanceled)).GetAwaiter().GetResult();`.
